components/gitpod-protocol/src/util/tracing.ts

```diff
@@ -186,6 +186,15 @@ export namespace TraceContext {
             context: owi,
         });
     }
+
+    export function finishOnce(span: opentracing.Span): () => void {
```

Comment:
Sounds like a hack on top of improper control flow.. but it will do

Reply:
Sometimes it's hard to know all the ways sth can behave, like in this case. Then the solution becomes a trade-off about where we want to spent our time. :shrug:

Reply:
Well, the span should always be opened on a middleware, and always closed on the middleware, irrespective of the outcome of individual functions.

In a way, we shouldn't even have to create the span at the start of the method, just add to a span which exists.

Reply:
>  the span should always be opened on a middleware, and always closed on the middleware

That's certainly sth we can aim for in the future. :+1: 

> In a way, we shouldn't even have to create the span at the start of the method, just add to a span which exists.

That's a question on granularity, which would also be interesting to discuss and align on. :100:

Reply:
This is one aspect where a service oriented architecture helps. Once the tracing is setup once, you get the same behaviour across all (baseserver) servers, and you get normalized tracing, even when you've not instrumented it explicitly. You can always add more child spans, when relevant.

Propagation of these, through context ends up being natural in Go.